### PR TITLE
refactor: charmed-kubeflow-chisme's base charm to re-write arg-controller

### PR DESCRIPTION
**Pull request by DnPlas**
This PR refactors the charm code using charmed-kubeflow-chisme's base charm for kubernetes resources reconciliation and event handling. In this PR are also included other changes needed to support this refactor:
* Update requirement input files
* Update libraries
* Update unit tests

TODO
- [ ] verify why in unit tests the assertion for applied resources passes with ==50 and not ==25 (the actual number of resources that should be applied)
- [x] debug error message  `[kubernetes:auth-crds-cm-and-secrets] Not all resources found in cluster.  This may be transient if we haven't tried ...` when deploying, could be related to https://github.com/canonical/charmed-kubeflow-chisme/issues/77
- [x] submit a PR in canonical/bundle-kubeflow to set `trust: True` in the bundle definition: https://github.com/canonical/bundle-kubeflow/pull/694

### CODE REVIEW

File: charms/argo-controller/src/charm.py [hunk 1]
```diff
@@ -54,260 +80,93 @@ def __init__(self, *args):
...
-        except NoCompatibleVersions as err:
-            raise CheckFailed(err, BlockedStatus)
-        return interfaces
-
-    def _check_object_storage(self, interfaces):
-        if not ((os := interfaces["object-storage"]) and os.get_data()):
-            raise CheckFailed("Waiting for object-storage relation data", BlockedStatus)
-
-        return list(os.get_data().values())[0]
+        self.profile_controller_container = self.charm_reconciler.add(
```
Comment (kimwnasptd) [inline charms/argo-controller/src/charm.py:154]: nit: the name shouldn't be `profile_controller_container` here

Comment (DnPlas) [reply charms/argo-controller/src/charm.py]: Updated.

**Comment by i-chvets** [comment]
@DnPlas What are requirements to run integration tests? Juju version? Juju agent version for controller?
Just running integration tests with Juju 2.9.44 fails with error: `juju.errors.JujuConnectionError: juju server-version 2.9.44 not supported`

The code looks good, reviewing.

File: charms/argo-controller/src/components/pebble_component.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+LIVENESS_PROBE_NAME = "argo-controller-up"
+
+
+class ArgoControllerPebbleService(PebbleServiceComponent):
+    def __init__(
+        self,
+        *args,
+        **kwargs,
+    ):
+        """Pebble service container component in order to configure Pebble layer"""
```
Comment (i-chvets) [inline charms/argo-controller/src/components/pebble_component.py:24]: May be this doc string should be on line 19?

File: charms/argo-controller/src/charm.py [hunk 3]
```diff
@@ -54,260 +80,93 @@ def __init__(self, *args):
...
-                    "updateStrategy": {
-                        "type": "RollingUpdate",
-                        "rollingUpdate": {"maxUnavailable": 1},
-                    },
+                krh_labels=create_charm_default_labels(
+                    self.app.name,
+                    self.model.name,
+                    scope="auth-crds-cm-and-secrets",
+                ),
+                context_callable=lambda: {
```
Comment (i-chvets) [inline charms/argo-controller/src/charm.py:122]: This inline JSON object makes it a bit hard to read. Is it possible to create `_generate_context()` helper that would return this object and use here it instead?
Not sure if it is possible. Just nitpicking.

**Comment by DnPlas** [comment]
> @DnPlas What are requirements to run integration tests? Juju version? Juju agent version for controller? Just running integration tests with Juju 2.9.44 fails with error: `juju.errors.JujuConnectionError: juju server-version 2.9.44 not supported`
> 
> The code looks good, reviewing.

Hey @i-chvets , regarding integration tests, right now the CI is running on juju 3.1/stable and microk8s 1.25-strict/stable, but I don't see a reason why this charm won't work with another juju version. From the error you are getting it sounds more like a local environment issue.

File: charms/argo-controller/src/charm.py (hunk 3 above)
Comment (DnPlas) [reply charms/argo-controller/src/charm.py]: Updated

File: charms/argo-controller/src/components/pebble_component.py (hunk 2 above)
Comment (DnPlas) [reply charms/argo-controller/src/components/pebble_component.py]: Updated

**Comment by i-chvets** [comment]
LGTM.

**Review by i-chvets — APPROVED** [review]
Approved.

**Pull request merged**